File: SmartInterface/UserControls/Bootstrap/Bootstrap_Menu.ascx.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.UI;
using System.Web.UI.WebControls;
using BusinessObject.Administration;
using BusinessObject.Administration.dstMenuTableAdapters;
using System.Runtime.InteropServices;

namespace SmartInterface.UserControls.Bootstrap
{
    public partial class Bootstrap_Menu : System.Web.UI.UserControl
    {
        public string strMenuText = "";

        protected void Page_Load(object sender, EventArgs e)
        {

        }

        public void BuildMenu(bool blnForItemAdmin, int intUserID = 0, int intActiveMenuID = 0)
        {
            strMenuText = "<ul class=\"sidebar-menu\">";
            strMenuText += "<li Class=\"header\">منوی اصلی</li>";
            int intAction = 2;
            if (blnForItemAdmin == true)
            {
                intAction = 1;
            }

            BusinessObject.Administration.dstMenuTableAdapters.spr_Menu_Parent_SelectTableAdapter tadpParentMenu = new spr_Menu_Parent_SelectTableAdapter();
            dstMenu.spr_Menu_Parent_SelectDataTable dtblParentMenu = null;
            dtblParentMenu = tadpParentMenu.GetData(intAction, intUserID);
            foreach (dstMenu.spr_Menu_Parent_SelectRow drwParentMenu in dtblParentMenu.Rows)
            #region for context
            {
                bool blnIsMenuOpen = false;
                if ((intActiveMenuID == drwParentMenu.ID))
                {
                    blnIsMenuOpen = true;
                }
                else
                {
                    spr_Menu_CheckSubMenuIsaChild_SelectTableAdapter tadpMenuCheckIsSubMenu = new spr_Menu_CheckSubMenuIsaChild_SelectTableAdapter();
                    dstMenu.spr_Menu_CheckSubMenuIsaChild_SelectDataTable dtblMenuCheckIsSubMenu = null;
                    dtblMenuCheckIsSubMenu = tadpMenuCheckIsSubMenu.GetData(drwParentMenu.ID, intActiveMenuID);
                    if (dtblMenuCheckIsSubMenu.Rows.Count > 0)
                    {
                        blnIsMenuOpen = true;
                    }

                }

                if (blnIsMenuOpen == true)
                {
                    strMenuText += "<li class=\"treeview active\">";
                }
                else
                {
                    strMenuText += "<li class=\"treeview\">";
                }




                if (drwParentMenu.IsLeaf == true && drwParentMenu.IsURLNull() == false)
                {
                    strMenuText += "<a href = \"" + drwParentMenu.URL + "\" >";
                }
                else
                {
                    strMenuText += "<a href = \"#\" >";
                }

                strMenuText += "<i class=\"" + drwParentMenu.IconStyle + "\"></i><span> " + drwParentMenu.MenuTitle + "</span>";
                if (drwParentMenu.IsLeaf == false)
                {
                    strMenuText += "<i class=\"fa fa-angle-left pull-right\"></i>";
                }
                else if (drwParentMenu.HasExtraInfo == true)
                {
                    strMenuText += "<small class=\"" + drwParentMenu.ExtraInfoStyle + "\">" + this.GetMenuExtraInfo(drwParentMenu.ID) + "</small>";
                }


                strMenuText += "</a>";
                strMenuText += this.GetMenuChildrenString(blnForItemAdmin, drwParentMenu.ID, intUserID, intActiveMenuID);
                strMenuText += "</li>";
            }
            #endregion
            strMenuText += "</ul>";
            
        }

        private string GetMenuChildrenString(bool blnForItemAdmin, int inttheParentID, int intUserID=0, int intActiveMenuID=0)
        {
          string strResultHTML = "";
          int intAction = 2;
            if (blnForItemAdmin == true)
            {
                intAction = 1;
            }
            
            strResultHTML = "<ul class=\"treeview-menu\">";
            spr_Menu_Child_SelectTableAdapter adtpChildMenu = new spr_Menu_Child_SelectTableAdapter();
            dstMenu.spr_Menu_Child_SelectDataTable dtblChildMenu = null;
            dtblChildMenu = adtpChildMenu.GetData(intAction, intUserID, inttheParentID);
            foreach (dstMenu.spr_Menu_Child_SelectRow drwChidMenu in dtblChildMenu.Rows)
            {
                bool  blnIsMenuOpen = false;
                if (intActiveMenuID == drwChidMenu.ID)
                {
                    blnIsMenuOpen = true;
                }
                else
                {
                    spr_Menu_CheckSubMenuIsaChild_SelectTableAdapter tadpMenuCheckIsSubMenu = new spr_Menu_CheckSubMenuIsaChild_SelectTableAdapter();
                    dstMenu.spr_Menu_CheckSubMenuIsaChild_SelectDataTable dtblMenuCheckIsSubMenu = null;
                    dtblMenuCheckIsSubMenu = tadpMenuCheckIsSubMenu.GetData(drwChidMenu.ID, intActiveMenuID);
                    if (dtblMenuCheckIsSubMenu.Rows.Count > 0)
                    {
                        blnIsMenuOpen = true;
                    }

                }

                if (blnIsMenuOpen == true)
                {
                    strResultHTML += "<li class=\"active\">";
                }
                else
                {
                    strResultHTML += "<li>";
                }

                if (drwChidMenu.IsLeaf == true && drwChidMenu.IsURLNull() == false)
                {
                    strResultHTML += "<a href = \"" + drwChidMenu.URL + "\" >";
                }
                else
                {
                    strResultHTML += "<a href = \"#\" >";
                }

                strResultHTML += "<i class=\"" + drwChidMenu.IconStyle + "\"></i><span> " + drwChidMenu.MenuTitle + "</span>";

                if ((drwChidMenu.IsLeaf == false))
                {
                    strResultHTML += "<i class=\"fa fa-angle-left pull-right\"></i>";
                }
                else if ((drwChidMenu.HasExtraInfo == true))
                {
                    strResultHTML += "<small class=\"" + drwChidMenu.ExtraInfoStyle + "\">" + this.GetMenuExtraInfo(drwChidMenu.ID) + "</small>";

                }

                strResultHTML += this.GetMenuChildrenString(blnForItemAdmin, drwChidMenu.ID, intUserID, intActiveMenuID);
                strResultHTML += "</a></li>";
            }

            strResultHTML += "</ul>";
            if (strResultHTML == "<ul class=\"treeview-menu\"></ul>")
            {
                return "";
            }
            else
            {
                return strResultHTML;
            }

        }

        private string GetMenuExtraInfo(int intMenuID)
        {
            string strExtraInfo = "";
            switch (intMenuID)
            {
                case 4:
                    strExtraInfo = "5";
                    break;
                case 12:
                    strExtraInfo = "new!";
                    break;
                case 17:
                    strExtraInfo = "24";
                    break;
            }
            return strExtraInfo;
        }






    }
}

File: RVAS/UserControls/Bootstrap/Bootstrap_Menu.ascx.vb
Imports BusinessObject.Administration

Public Class Bootstrap_Menu
    Inherits System.Web.UI.UserControl
    Public strMenuText As String = ""
    Protected Sub Page_Load(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.Load

    End Sub

    Public Sub BuildMenu(blnForItemAdmin As Boolean, Optional intUserID As Integer = 0, Optional intActiveMenuID As Integer = 0)
        strMenuText = "<ul class=""sidebar-menu"">"
        strMenuText &= "<li class=""header"">منوی اصلی</li>"
        Dim intAction As Integer = 2

        If blnForItemAdmin = True Then
            intAction = 1
        End If

        Dim tadpParentMenu As New BusinessObject.Administration.dstMenuTableAdapters.spr_Menu_Parent_SelectTableAdapter
        Dim dtblParentMenu As dstMenu.spr_Menu_Parent_SelectDataTable = Nothing
        dtblParentMenu = tadpParentMenu.GetData(intAction, intUserID)
        For Each drwParentMenu As dstMenu.spr_Menu_Parent_SelectRow In dtblParentMenu.Rows
#Region "For drwParentMenu"
            Dim blnIsMenuOpen As Boolean = False
            If intActiveMenuID = drwParentMenu.ID Then
                blnIsMenuOpen = True
            Else
                Dim tadpMenuCheckIsSubMenu As New dstMenuTableAdapters.spr_Menu_CheckSubMenuIsaChild_SelectTableAdapter()
                Dim dtblMenuCheckIsSubMenu As dstMenu.spr_Menu_CheckSubMenuIsaChild_SelectDataTable = Nothing
                dtblMenuCheckIsSubMenu = tadpMenuCheckIsSubMenu.GetData(drwParentMenu.ID, intActiveMenuID)
                If dtblMenuCheckIsSubMenu.Rows.Count > 0 Then
                    blnIsMenuOpen = True
                End If
            End If

            If blnIsMenuOpen = True Then
                strMenuText &= "<li class=""treeview active"">"
            Else
                strMenuText &= "<li class=""treeview"">"
            End If


            If drwParentMenu.IsLeaf = True AndAlso drwParentMenu.IsURLNull() = False Then
                strMenuText &= "<a href =""" & drwParentMenu.URL & """>"
            Else
                strMenuText &= "<a href=""#"" >"
            End If
            strMenuText &= "<i class=""" & drwParentMenu.IconStyle & """></i><span> " & drwParentMenu.MenuTitle & "</span>"

            If drwParentMenu.IsLeaf = False Then
                strMenuText &= "<i class=""fa fa-angle-left pull-right""></i>"
            ElseIf drwParentMenu.HasExtraInfo = True
                strMenuText &= "<small class=""" & drwParentMenu.ExtraInfoStyle & """>" & Me.GetMenuExtraInfo(drwParentMenu.ID) & "</small>"
            End If

            strMenuText &= "</a>"
            strMenuText &= Me.GetMenuChildrenString(blnForItemAdmin, drwParentMenu.ID, intUserID, intActiveMenuID)
            strMenuText &= "</li>"

#End Region
        Next drwParentMenu
        strMenuText &= "</ul>"
    End Sub



    Private Function GetMenuChildrenString(blnForItemAdmin As Boolean, inttheParentID As Integer, Optional intUserID As Integer = 0, Optional intActiveMenuID As Integer = 0) As String
        Dim strResultHTML As String = ""
        Dim intAction As Integer = 2
        If blnForItemAdmin = True Then
            intAction = 1
        End If

        strResultHTML = "<ul class=""treeview-menu"">"
        Dim adtpChildMenu As New dstMenuTableAdapters.spr_Menu_Child_SelectTableAdapter()
        Dim dtblChildMenu As dstMenu.spr_Menu_Child_SelectDataTable = Nothing
        dtblChildMenu = adtpChildMenu.GetData(intAction, intUserID, inttheParentID)
        For Each drwChidMenu As dstMenu.spr_Menu_Child_SelectRow In dtblChildMenu.Rows
            Dim blnIsMenuOpen As Boolean = False
            If intActiveMenuID = drwChidMenu.ID Then
                blnIsMenuOpen = True
            Else
                Dim tadpMenuCheckIsSubMenu As New dstMenuTableAdapters.spr_Menu_CheckSubMenuIsaChild_SelectTableAdapter()
                Dim dtblMenuCheckIsSubMenu As dstMenu.spr_Menu_CheckSubMenuIsaChild_SelectDataTable = Nothing
                dtblMenuCheckIsSubMenu = tadpMenuCheckIsSubMenu.GetData(drwChidMenu.ID, intActiveMenuID)
                If dtblMenuCheckIsSubMenu.Rows.Count > 0 Then
                    blnIsMenuOpen = True

                End If
            End If

            If blnIsMenuOpen = True Then
                strResultHTML &= "<li class=""active"">"
            Else
                strResultHTML &= "<li>"
            End If

            If drwChidMenu.IsLeaf = True AndAlso drwChidMenu.IsURLNull() = False Then
                strResultHTML &= "<a href= """ & drwChidMenu.URL & """>"
            Else
                strResultHTML &= "<a href=""#"">"
            End If

            strResultHTML &= ("<i class=""" & drwChidMenu.IconStyle & """></i><span> ") & drwChidMenu.MenuTitle & "</span>"

            If drwChidMenu.IsLeaf = False Then
                strResultHTML &= "<i class=""fa fa-angle-left pull-right""></i>"
            ElseIf drwChidMenu.HasExtraInfo = True Then

                strResultHTML &= "<small class=""" & drwChidMenu.ExtraInfoStyle & """>" & Me.GetMenuExtraInfo(drwChidMenu.ID) & "</small>"
            End If
            strResultHTML &= "</a>"
            strResultHTML &= Me.GetMenuChildrenString(blnForItemAdmin, drwChidMenu.ID, intUserID, intActiveMenuID)
            strResultHTML &= "</li>"
        Next

        strResultHTML &= "</ul>"
        If strResultHTML = "<ul class=""treeview-menu""></ul>" Then
            Return ""
        Else
            Return strResultHTML
        End If

    End Function

    Private Function GetMenuExtraInfo(intMenuID As Integer) As String
        Dim strExtraInfo As String = ""
        Select Case intMenuID
            Case 4
                strExtraInfo = "5"
                Exit Select
            Case 12
                strExtraInfo = "new!"
                Exit Select
            Case 17
                strExtraInfo = "24"
                Exit Select
        End Select
        Return strExtraInfo
    End Function

End Class